File: test/TEST_ArgParser.cs
using System;
using System.Collections;
using System.Collections.Generic;

using Xunit;

using InoCLI;

namespace InoCLI.TEST
{
   // ============================================================
   /// <summary>
   /// Tests for ArgParser: positionals, optionals, helpers.
   /// </summary>
   // ============================================================
   public class TEST_ArgParser
   {

   #region Positionals

      [Fact]
      public void Parse_SinglePositional()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Single(result.Positionals);
         Assert.Equal("ping", result[0]);
      }

      [Fact]
      public void Parse_MultiplePositionals()
      {
         var result = new ArgParser().Parse(new[] { "build", "src/main.cs", "42" });

         Assert.Equal(3, result.Positionals.Count);
         Assert.Equal("build", result[0]);
         Assert.Equal("src/main.cs", result[1]);
         Assert.Equal("42", result[2]);
      }

      [Fact]
      public void Parse_NegativeNumber_IsPositional()
      {
         var result = new ArgParser().Parse(new[] { "calc", "-42" });

         Assert.Equal(2, result.Positionals.Count);
         Assert.Equal("-42", result[1]);
      }

   #endregion

   #region Optionals — Long

      [Fact]
      public void Parse_LongOption_WithValue()
      {
         var result = new ArgParser().Parse(new[] { "deploy", "--retries", "3" });

         Assert.True(result.Has("retries"));
         Assert.Equal("3", result["retries"]);
      }

      [Fact]
      public void Parse_LongOption_Flag()
      {
         var result = new ArgParser().Parse(new[] { "status", "--verbose" });

         Assert.True(result.Has("verbose"));
         Assert.Empty(result.Optionals["verbose"]);
      }

      [Fact]
      public void Parse_LongOption_Repeated()
      {
         var result = new ArgParser().Parse(new[] { "build", "src/main.cs", "--tag", "release", "--tag", "latest" });

         var tags = result.Optionals["tag"];

         Assert.Equal(2, tags.Count);
         Assert.Equal("release", tags[0]);
         Assert.Equal("latest", tags[1]);
      }

   #endregion

   #region Optionals — Short

      [Fact]
      public void Parse_ShortOption_WithValue()
      {
         var result = new ArgParser().Parse(new[] { "deploy", "-r", "3" });

         Assert.True(result.Has("r"));
         Assert.Equal("3", result["r"]);
      }

      [Fact]
      public void Parse_ShortOption_Flag()
      {
         var result = new ArgParser().Parse(new[] { "status", "-v" });

         Assert.True(result.Has("v"));
         Assert.Empty(result.Optionals["v"]);
      }

   #endregion

   #region Mixed

      [Fact]
      public void Parse_OptionsBetweenPositionals()
      {
         var result = new ArgParser().Parse(new[] { "build", "src/main.cs", "--filter", "x > 0", "42" });

         Assert.Equal("build", result[0]);
         Assert.Equal("src/main.cs", result[1]);
         Assert.Equal("42", result[2]);
         Assert.Equal("x > 0", result["filter"]);
      }

      [Fact]
      public void Parse_Empty()
      {
         var result = new ArgParser().Parse(new string[0]);

         Assert.Empty(result.Positionals);
         Assert.Empty(result.Optionals);
      }

   #endregion

   #region Error

      [Fact]
      public void Parse_DoubleDash_Throws()
      {
         Assert.Throws<ArgumentException>(() => new ArgParser().Parse(new[] { "--" }));
      }

   #endregion

   #region Indexer

      [Fact]
      public void Indexer_Positional_ReturnsValue()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Equal("ping", result[0]);
      }

      [Fact]
      public void Indexer_Positional_OutOfRange_ReturnsNull()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Null(result[5]);
      }

      [Fact]
      public void Indexer_Optional_ReturnsFirstValue()
      {
         var result = new ArgParser().Parse(new[] { "--retries", "3" });

         Assert.Equal("3", result["retries"]);
      }

      [Fact]
      public void Indexer_Optional_Missing_ReturnsNull()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Null(result["missing"]);
      }

   #endregion

   #region Has

      [Fact]
      public void Has_Positional_Exists()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.True(result.Has(0));
      }

      [Fact]
      public void Has_Positional_Missing()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.False(result.Has(1));
      }

      [Fact]
      public void Has_Optional_Exists()
      {
         var result = new ArgParser().Parse(new[] { "--verbose" });

         Assert.True(result.Has("verbose"));
      }

      [Fact]
      public void Has_Optional_Missing()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.False(result.Has("missing"));
      }

   #endregion

   #region GetInt

      [Fact]
      public void GetInt_Positional()
      {
         var result = new ArgParser().Parse(new[] { "connect", "8080" });

         Assert.Equal(8080, result.GetInt(1));
      }

      [Fact]
      public void GetInt_Positional_Missing_Throws()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Throws<ArgumentException>(() => result.GetInt(5));
      }

      [Fact]
      public void GetInt_Optional()
      {
         var result = new ArgParser().Parse(new[] { "--retries", "3" });

         Assert.Equal(3, result.GetInt("retries"));
      }

      [Fact]
      public void GetInt_Optional_Missing_Throws()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Throws<ArgumentException>(() => result.GetInt("retries"));
      }

      [Fact]
      public void GetInt_Optional_Invalid_Throws()
      {
         var result = new ArgParser().Parse(new[] { "--retries", "abc" });

         Assert.Throws<ArgumentException>(() => result.GetInt("retries"));
      }

   #endregion

   #region GetFloat

      [Fact]
      public void GetFloat_Positional()
      {
         var result = new ArgParser().Parse(new[] { "resize", "1.5" });

         Assert.Equal(1.5f, result.GetFloat(1));
      }

      [Fact]
      public void GetFloat_Optional()
      {
         var result = new ArgParser().Parse(new[] { "--ratio", "1.5" });

         Assert.Equal(1.5f, result.GetFloat("ratio"));
      }

   #endregion

   #region GetBool

      [Fact]
      public void GetBool_Positional()
      {
         var result = new ArgParser().Parse(new[] { "toggle", "true" });

         Assert.True(result.GetBool(1));
      }

      [Fact]
      public void GetBool_Optional()
      {
         var result = new ArgParser().Parse(new[] { "--verbose", "true" });

         Assert.True(result.GetBool("verbose"));
      }

   #endregion

   #region All

      [Fact]
      public void All_ReturnsValues()
      {
         var result = new ArgParser().Parse(new[] { "--tag", "release", "--tag", "latest" });

         var values = result.All("tag");

         Assert.Equal(2, values.Count);
         Assert.Equal("release", values[0]);
         Assert.Equal("latest", values[1]);
      }

      [Fact]
      public void All_Missing_Throws()
      {
         var result = new ArgParser().Parse(new[] { "ping" });

         Assert.Throws<ArgumentException>(() => result.All("missing"));
      }

   #endregion

   }
}

File: rust/inocli/tests/test_arg_parser.rs
use inocli::{ArgParser, CliError};

fn parse(args: &[&str]) -> inocli::CommandArgs {
    ArgParser::new().parse(args).unwrap()
}

// ================================================================
// Positionals
// ================================================================

#[test]
fn parse_single_positional() {
    let result = parse(&["ping"]);
    assert_eq!(result.positionals.len(), 1);
    assert_eq!(result.get(0), Some("ping"));
}

#[test]
fn parse_multiple_positionals() {
    let result = parse(&["build", "src/main.rs", "42"]);
    assert_eq!(result.count(), 3);
    assert_eq!(result.get(0), Some("build"));
    assert_eq!(result.get(1), Some("src/main.rs"));
    assert_eq!(result.get(2), Some("42"));
}

#[test]
fn parse_negative_number_is_positional() {
    let result = parse(&["calc", "-42"]);
    assert_eq!(result.count(), 2);
    assert_eq!(result.get(1), Some("-42"));
}

// ================================================================
// Long options
// ================================================================

#[test]
fn parse_long_option_with_value() {
    let result = parse(&["deploy", "--retries", "3"]);
    assert!(result.has_opt("retries"));
    assert_eq!(result.opt("retries"), Some("3"));
}

#[test]
fn parse_long_option_flag() {
    let result = parse(&["status", "--verbose"]);
    assert!(result.has_opt("verbose"));
    assert!(result.optionals["verbose"].is_empty());
}

#[test]
fn parse_long_option_repeated() {
    let result = parse(&["build", "src/main.rs", "--tag", "release", "--tag", "latest"]);
    let tags = &result.optionals["tag"];
    assert_eq!(tags.len(), 2);
    assert_eq!(tags[0], "release");
    assert_eq!(tags[1], "latest");
}

// ================================================================
// Short options
// ================================================================

#[test]
fn parse_short_option_with_value() {
    let result = parse(&["deploy", "-r", "3"]);
    assert!(result.has_opt("r"));
    assert_eq!(result.opt("r"), Some("3"));
}

#[test]
fn parse_short_option_flag() {
    let result = parse(&["status", "-v"]);
    assert!(result.has_opt("v"));
    assert!(result.optionals["v"].is_empty());
}

// ================================================================
// Mixed
// ================================================================

#[test]
fn parse_options_between_positionals() {
    let result = parse(&["build", "src/main.rs", "--filter", "x > 0", "42"]);
    assert_eq!(result.get(0), Some("build"));
    assert_eq!(result.get(1), Some("src/main.rs"));
    assert_eq!(result.get(2), Some("42"));
    assert_eq!(result.opt("filter"), Some("x > 0"));
}

#[test]
fn parse_empty() {
    let result = parse(&[]);
    assert!(result.positionals.is_empty());
    assert!(result.optionals.is_empty());
}

// ================================================================
// Error
// ================================================================

#[test]
fn parse_double_dash_errors() {
    let err = ArgParser::new().parse(&["--"]).unwrap_err();
    assert!(matches!(err, CliError::InvalidOption(_)));
}

// ================================================================
// Indexer / get
// ================================================================

#[test]
fn get_positional_returns_value() {
    let result = parse(&["ping"]);
    assert_eq!(result.get(0), Some("ping"));
}

#[test]
fn get_positional_out_of_range_returns_none() {
    let result = parse(&["ping"]);
    assert_eq!(result.get(5), None);
}

#[test]
fn opt_returns_first_value() {
    let result = parse(&["--retries", "3"]);
    assert_eq!(result.opt("retries"), Some("3"));
}

#[test]
fn opt_missing_returns_none() {
    let result = parse(&["ping"]);
    assert_eq!(result.opt("missing"), None);
}

// ================================================================
// Has
// ================================================================

#[test]
fn has_positional_exists() {
    let result = parse(&["ping"]);
    assert!(result.has(0));
}

#[test]
fn has_positional_missing() {
    let result = parse(&["ping"]);
    assert!(!result.has(1));
}

#[test]
fn has_opt_exists() {
    let result = parse(&["--verbose"]);
    assert!(result.has_opt("verbose"));
}

#[test]
fn has_opt_missing() {
    let result = parse(&["ping"]);
    assert!(!result.has_opt("missing"));
}

// ================================================================
// get_int
// ================================================================

#[test]
fn get_int_positional() {
    let result = parse(&["connect", "8080"]);
    assert_eq!(result.get_int(1).unwrap(), 8080);
}

#[test]
fn get_int_positional_missing_errors() {
    let result = parse(&["ping"]);
    assert!(matches!(result.get_int(5), Err(CliError::MissingPositional(5))));
}

#[test]
fn get_int_optional() {
    let result = parse(&["--retries", "3"]);
    assert_eq!(result.get_int_opt("retries").unwrap(), 3);
}

#[test]
fn get_int_optional_missing_errors() {
    let result = parse(&["ping"]);
    assert!(matches!(result.get_int_opt("retries"), Err(CliError::MissingOption(_))));
}

#[test]
fn get_int_optional_invalid_errors() {
    let result = parse(&["--retries", "abc"]);
    assert!(matches!(result.get_int_opt("retries"), Err(CliError::ParseError { .. })));
}

// ================================================================
// get_f32
// ================================================================

#[test]
fn get_f32_positional() {
    let result = parse(&["resize", "1.5"]);
    assert!((result.get_f32(1).unwrap() - 1.5f32).abs() < f32::EPSILON);
}

#[test]
fn get_f32_optional() {
    let result = parse(&["--ratio", "1.5"]);
    assert!((result.get_f32_opt("ratio").unwrap() - 1.5f32).abs() < f32::EPSILON);
}

// ================================================================
// get_bool
// ================================================================

#[test]
fn get_bool_positional() {
    let result = parse(&["toggle", "true"]);
    assert!(result.get_bool(1).unwrap());
}

#[test]
fn get_bool_optional() {
    let result = parse(&["--verbose", "true"]);
    assert!(result.get_bool_opt("verbose").unwrap());
}

#[test]
fn get_bool_case_insensitive() {
    let result = parse(&["toggle", "True"]);
    assert!(result.get_bool(1).unwrap());
}

// ================================================================
// all
// ================================================================

#[test]
fn all_returns_values() {
    let result = parse(&["--tag", "release", "--tag", "latest"]);
    let values = result.all("tag").unwrap();
    assert_eq!(values.len(), 2);
    assert_eq!(values[0], "release");
    assert_eq!(values[1], "latest");
}

#[test]
fn all_missing_errors() {
    let result = parse(&["ping"]);
    assert!(matches!(result.all("missing"), Err(CliError::MissingOption(_))));
}

// ================================================================
// CommandArgs helpers (mirrors TEST_CommandRegistry helpers section)
// ================================================================

#[test]
fn flag_returns_correctly() {
    let mut args = inocli::CommandArgs::new();
    args.optionals.insert("verbose".to_string(), vec![]);
    assert!(args.flag("verbose"));
    assert!(!args.flag("missing"));
}

#[test]
fn get_int_throws_and_fallback() {
    let mut args = inocli::CommandArgs::new();
    args.positionals = vec!["42".to_string(), "abc".to_string()];
    args.optionals.insert("count".to_string(), vec!["10".to_string()]);

    assert_eq!(args.get_int(0).unwrap(), 42);
    assert!(args.get_int(1).is_err());

    assert_eq!(args.get_int_or(0, 0), 42);
    assert_eq!(args.get_int_or(1, 99), 99);
    assert_eq!(args.get_int_or(5, 99), 99);

    assert_eq!(args.get_int_opt_or("count", 0), 10);
    assert_eq!(args.get_int_opt_or("missing", 99), 99);
}

#[test]
fn get_f32_throws_and_fallback() {
    let mut args = inocli::CommandArgs::new();
    args.positionals = vec!["3.14".to_string()];
    args.optionals.insert("ratio".to_string(), vec!["0.5".to_string()]);

    assert!((args.get_f32(0).unwrap() - 3.14f32).abs() < 0.001);
    assert!((args.get_f32_opt("ratio").unwrap() - 0.5f32).abs() < f32::EPSILON);
    assert!((args.get_f32_or(5, 1.0f32) - 1.0f32).abs() < f32::EPSILON);
    assert!((args.get_f32_opt_or("missing", 1.0f32) - 1.0f32).abs() < f32::EPSILON);
}

#[test]
fn get_f64_throws_and_fallback() {
    let mut args = inocli::CommandArgs::new();
    args.positionals = vec!["3.141592653589793".to_string()];
    args.optionals.insert("precision".to_string(), vec!["0.001".to_string()]);

    assert!((args.get_f64(0).unwrap() - std::f64::consts::PI).abs() < 1e-10);
    assert!((args.get_f64_opt("precision").unwrap() - 0.001).abs() < 1e-10);
    assert_eq!(args.get_f64_or(5, 1.0), 1.0);
    assert_eq!(args.get_f64_opt_or("missing", 1.0), 1.0);
}

#[test]
fn get_bool_throws_and_fallback() {
    let mut args = inocli::CommandArgs::new();
    args.positionals = vec!["true".to_string()];
    args.optionals.insert("enabled".to_string(), vec!["false".to_string()]);

    assert!(args.get_bool(0).unwrap());
    assert!(!args.get_bool_opt("enabled").unwrap());
    assert!(args.get_bool_or(5, true));
    assert!(args.get_bool_opt_or("missing", true));
}

#[test]
fn from_index_returns_slice() {
    let mut args = inocli::CommandArgs::new();
    args.positionals = vec!["a".to_string(), "b".to_string(), "c".to_string(), "d".to_string()];

    assert_eq!(args.from_index(2), &["c".to_string(), "d".to_string()]);
    assert_eq!(args.from_index(0), &["a".to_string(), "b".to_string(), "c".to_string(), "d".to_string()]);
    assert!(args.from_index(10).is_empty());
}

#[test]
fn all_throws_and_fallback() {
    let mut args = inocli::CommandArgs::new();
    args.optionals.insert("tag".to_string(), vec!["a".to_string(), "b".to_string(), "c".to_string()]);

    assert_eq!(args.all("tag").unwrap(), &["a".to_string(), "b".to_string(), "c".to_string()]);
    assert!(matches!(args.all("missing"), Err(CliError::MissingOption(_))));

    let fallback = vec!["default".to_string()];
    assert_eq!(args.all_or("missing", &fallback), fallback.as_slice());
}
